docs: rework tensor operations report (#74)
Shad00Z/machine-learning-compilers: docs/submissions/05_tensor_op.rst

5.5.1.3 Ternary Operation
"""""""""""""""""""""""""""""""""

We first check for a second ``K`` dimension, which has no unit stride in the second input tensor and does not appear in the output tensor.
If such a ``K`` dimension exists, we select it as the batch-reduce dimension and conclude that the current operation is a BRGEMM.
If no such ``K`` dimension exists, we are handling a GEMM.

Next, we identify the primitive ``M`` dimension by checking for an ``M`` dimension with unit stride in the first input and output tensor.

The primary ``N`` dimension is then found by choosing the dimension with the smallest stride which appears only in the second input tensor and in the output tensor.

Lastly, we identify the primitive ``K`` dimension by searching for a ``K`` dimension that has unit stride in the second input tensor and does not appear in the output tensor.

5.5.2 Dimension Splitting
^^^^^^^^^^^^^^^^^^^^^^^^^^^

For our second optimization pass we decided to look at the dimension sizes of our loops. 
We introduced a ``max_kernel_size`` parameter, which specifies the maximum allowed size for a dimension.

> If a dimension with a size larger than the maximum is found, the dimension splitter will try to split it into new dimensions with optimized sizes. 

Reviewer: ```suggestion
If a dimension size exceeds the maximum size, the dimension splitter will try to split it into new dimensions with optimized sizes. 
```
Personal preference for readability

Depending on the selected dimensions our results varied massively. The highest performance we achieved was around ``350 GFLOPs``. 

.. _unary-operations:

**********************************
5.6 Unary Operations
**********************************

5.6.1 Backend Extension
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In this task, we were supposed to add support for unary operations, such as permuting a tensor's dimensions, to our tensor operation backend.
Furthermore, we had to implement primitive identification and shared memory parallelization optimization passes for these unary primitives.

Instead of separating the documentation on unary, binary and ternary operations, we decided to merge them.
This means that the code for handling unary operations has already been shown and explained in the previous section(s).

5.6.2 Reference Implementation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> For our reference implementation, we used an example with 4 dimensions ``trus`` in which we reorder the dimensions to ``turs``.

Reviewer: ```suggestion
For our reference implementation, we used an example with 4 dimensions, ``trus``, where we reorder the dimensions to ``turs``.
```
Again, personal preference. Yours works as well
